llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeWinogradInputPass.cpp

```diff
@@ -0,0 +1,246 @@
+
+    rewriter.setInsertionPointToStart(&funcOp.getBody().front());
+    auto zeroF32 = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(elementType));
+    auto scratch =
+        rewriter.create<tensor::EmptyOp>(loc, inputTileSquare, elementType);
+
+    rewriter.setInsertionPoint(inputOp);
+    SmallVector<Value> lbs, ubs, steps;
+    computeLoopParams(lbs, ubs, steps, output, loc, rewriter);
+    // Construct loops
+    auto loopNest = scf::buildLoopNest(
+        rewriter, loc, lbs, ubs, steps, ValueRange({output}),
+        [&](OpBuilder &nestedBuilder, Location loc, ValueRange outputIvs,
+            ValueRange iterArgs) -> scf::ValueVector { return {iterArgs[0]}; });
+
+    // Extract input slice
+    auto one = rewriter.getIndexAttr(1);
+    auto zero = rewriter.getIndexAttr(0);
+    auto inputTileSizeAttr = rewriter.getIndexAttr(inputTileSize);
+    SmallVector<OpFoldResult> strides(inputOp.getInputOperandRank(), one);
+    SmallVector<OpFoldResult> sizes(inputOp.getInputOperandRank(), one);
+    SmallVector<OpFoldResult> offsets(inputOp.getInputOperandRank(), zero);
+    SmallVector<Value> ivs;
+    for (auto loop : loopNest.loops) {
+      ivs.push_back(loop.getInductionVar());
+    }
+    for (int i = 0; i < inputShape.size(); i++) {
+      if ((i == 0) || (i == 3)) {
+        offsets[i] = ivs[i];
+      } else {
+        rewriter.setInsertionPointToStart(loopNest.loops[i].getBody());
+        AffineExpr dim0;
+        auto it = rewriter.getAffineConstantExpr(inputTileSize);
+        auto ot = rewriter.getAffineConstantExpr(outputTileSize);
+        auto delta = rewriter.getAffineConstantExpr(inputShape[i]);
+        bindDims(rewriter.getContext(), dim0);
+        AffineMap scaleMap =
+            AffineMap::get(1, 0, {dim0 * ot}, rewriter.getContext());
+        offsets[i] = rewriter.createOrFold<AffineApplyOp>(loc, scaleMap,
+                                                          ValueRange{ivs[i]});
+        AffineMap minMap =
+            AffineMap::get(1, 0, {-dim0 + delta, it}, rewriter.getContext());
+        sizes[i] = rewriter.createOrFold<AffineMinOp>(
+            loc, minMap,
+            ValueRange{
+                getValueOrCreateConstantIndexOp(rewriter, loc, offsets[i])});
+      }
+    }
+    rewriter.setInsertionPointToStart(loopNest.loops.back().getBody());
+    auto tensorType = RankedTensorType::get(
+        SmallVector<int64_t>(2, ShapedType::kDynamicSize), elementType);
+    auto dynamicSlice = rewriter
+                            .create<tensor::ExtractSliceOp>(
+                                loc, tensorType, input, offsets, sizes, strides)
+                            .getResult();
+
+    // Copy input slice into zeroed padded scratch space
+    strides = SmallVector<OpFoldResult>(2, one);
+    offsets = SmallVector<OpFoldResult>(2, zero);
+    sizes = SmallVector<OpFoldResult>{sizes[1], sizes[2]};
+    Value zeroScratch = rewriter
```

Comment:
I think we agreed to use `tensor.pad` if that's possible?

Reply:
I was able to get my original version to work (after fixing the issue related to using iter args) and since all my initial results were based on this form, I decided to keep this as the baseline and was planning to add tensor.pad as a performance optimization in a subsequent patch.

Reply:
SG!